gtn/functions/compose.cpp

```diff
@@ -450,3 +456,940 @@ Graph compose(
+        ? inArcOffset - graphDP1.inArcOffset[node]
+        : outArcOffset - graphDP1.outArcOffset[node];
+
+    node = toExploreNodePair[i].second;
+    // Special case if it is the last node. Then the offset becomes
+    // the number of arcs
+    inArcOffset = ((node + 1) == graphDP2.inArcOffset.size())
+        ? graphDP2.inArcs.size()
+        : graphDP2.inArcOffset[node + 1];
+    outArcOffset = ((node + 1) == graphDP2.outArcOffset.size())
+        ? graphDP2.outArcs.size()
+        : graphDP2.outArcOffset[node + 1];
+
+    const int numArcsSecond = inOrOutArc
+        ? inArcOffset - graphDP2.inArcOffset[node]
+        : outArcOffset - graphDP2.outArcOffset[node];
+
+    toExploreNumArcs[i] = std::make_pair(numArcsFirst, numArcsSecond);
+    arcCrossProductOffset[i] = numArcsFirst * numArcsSecond;
+  }
+
+  return std::make_tuple(arcCrossProductOffset, toExploreNumArcs);
+}
+
+// This function needs to be thread safe since multiple threads can
+// can call it and they will overlap on curIdx and dstIdx
+void calculateNumArcsAndNodesToExplore(
+    int curIdx,
+    int dstIdx,
+    const std::vector<bool>& reachable,
+    std::vector<bool>& newNodes,
+    std::vector<bool>& toExplore,
+    std::vector<int>& numOutArcs,
+    std::vector<int>& numInArcs) {
+  if (reachable[dstIdx]) {
+    // Atomic test and set for newNodes
+    if (!newNodes[dstIdx]) {
+      newNodes[dstIdx] = true;
+      toExplore[dstIdx] = true;
+    }
+
+    // These are atomic increments
+    numOutArcs[curIdx]++;
+    numInArcs[dstIdx]++;
+  }
+}
+
+// This function needs to be thread safe since multiple threads can
+// can call it
+void generateCombinedGraphNodesAndArcs(
+    int dstIdx,
+    int curIdx,
+    const std::pair<int, int>& arcPair,
+    const std::vector<bool>& reachable,
+    const std::vector<int>& newNodesOffset,
+    std::vector<bool>& newNodesVisited,
+    std::vector<bool>& toExplore,
+    std::pair<std::vector<int>, std::vector<int>>& gradInfo,
+    GraphDataParallel& newGraphDP,
+    std::pair<bool, bool> srcNodeStartAndAccept,
+    std::pair<bool, bool> dstNodeStartAndAccept,
+    int ilabel,
+    int olabel,
+    int weight) {
+  if (reachable[dstIdx]) {
+    // Atomic test and set for newNodesVisited
+    if (!newNodesVisited[dstIdx]) {
+      newNodesVisited[dstIdx] = true;
+      toExplore[dstIdx] = true;
+    }
+
+    // Set accept and start nodes
+    // I think I only need it for dst nodes and src nodes
+    // Note: Multiple threads can have the same dstIdx and write to the same
+    //       location and collide. This _should_ be fine since they are going
+    //       to write the same value
+    newGraphDP.start[newNodesOffset[dstIdx]] = dstNodeStartAndAccept.first;
+    newGraphDP.accept[newNodesOffset[dstIdx]] = dstNodeStartAndAccept.second;
+
+    // Both of these increments are atomic
+    int inArcIdx = newGraphDP.inArcOffset[newNodesOffset[dstIdx]]++;
+    int outArcIdx = newGraphDP.outArcOffset[newNodesOffset[curIdx]]++;
```

Comment:
I think the `++` here are a bug? The `inArcOffset` and `outArcOffset` arrays have already been scanned at this point.